internal/sync/sync.go

```diff
@@ -114,10 +117,16 @@ func (s *Synchronizer) Run() {
 
 func (s *Synchronizer) handleSync() {
 	for {
-		err := s.sync()
-		s.logger.With("Error", err).Info("Sync Failed, restarting iterator in 10 seconds")
-		time.Sleep(10 * time.Second)
-		s.stateDiffCollector.Close()
+		select {
+		case <-s.quit:
+			return
+		default:
+			err := s.sync()
+			s.logger.With("Error", err).Info("Sync Failed, restarting iterator in 10 seconds")
+			time.Sleep(10 * time.Second)
+			s.stateDiffCollector.Close()
+
+		}
 	}
 }
```

Comment:
I think this loop always gets stuck on the first iteration because line 124 (`s.sync()`) starts the sync process which always waits for a new state update [here](https://github.com/NethermindEth/juno/blob/f2b29bd0fc9212093f95e5e4122aeb4c127fe158/internal/sync/sync.go#L140) unless an error occurs. So, that means for me the race condition is still happening inside the `sync` function

Reply:
[This line](https://github.com/NethermindEth/juno/blob/f2b29bd0fc9212093f95e5e4122aeb4c127fe158/internal/sync/sync.go#L140) will also exit when the channel returned  `s.stateDiffCollector.GetChannel()` is closed. The for loop over a channel block forever unless it is closed. Once a channel is closed the for loop exits. Which then in turn causes `sync()` to exit.

The channel returned from `s.stateDiffCollector.GetChannel()` is closed when `s.stateDiffCollector.Close()` is called in `synchroniser.Close()` . Therefore, when the `sync()` returns without an error it will go over the loop again and then it will see that `s.quit` is closed and hence `handleSync()` would exit.

Reply:
Please look at the latest commit